symengine/numer_denom.cpp

```diff
@@ -0,0 +1,130 @@
+#include <symengine/basic.h>
+#include <symengine/symbol.h>
+#include <symengine/add.h>
+#include <symengine/integer.h>
+#include <symengine/rational.h>
+#include <symengine/complex.h>
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/functions.h>
+#include <symengine/constants.h>
+#include <symengine/visitor.h>
+
+namespace SymEngine {
+
+class NumerDenomVisitor : public BaseVisitor<NumerDenomVisitor> {
+private:
+    Ptr<RCP<const Basic>> numer_, denom_;
+public:
+    NumerDenomVisitor(const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom)
+        : numer_{numer}, denom_{denom} { }
+
+    void apply(const Basic &b) {
+        b.accept(*this);
+    }
+
+    void bvisit(const Mul &x) {
+
+        RCP<const Basic> curr_num = one;
+        RCP<const Basic> curr_den = one;
+        RCP<const Basic> arg_num, arg_den;
+
+        for (const auto &arg: x.get_args()) {
+            as_numer_denom(arg, outArg(arg_num), outArg(arg_den));
+            curr_num = mul(curr_num, arg_num);
+            curr_den = mul(curr_den, arg_den);
+        }
+
+        *numer_ = curr_num;
+        *denom_ = curr_den;
+    }
+
+    void bvisit(const Add &x) {
+
+        RCP<const Basic> curr_num = zero;
+        RCP<const Basic> curr_den = one;
+        RCP<const Basic> arg_num, arg_den, den_mul, divx;
+        RCP<const Basic> divx_num, divx_den;
+
+        for (const auto &arg: x.get_args()) {
+            as_numer_denom(arg, outArg(arg_num), outArg(arg_den));
+
+            divx = div(arg_den, curr_den);
+            as_numer_denom(divx, outArg(divx_num), outArg(divx_den));
+            if (eq(*divx_den, *one)) {
+                // the curr_den completely divides the arg_den
+                // can be made more extensive
+                curr_den = arg_den;
+                curr_num = add(mul(curr_num, divx), arg_num);
+                continue;
+            }
+
+            divx = div(curr_den, arg_den);
+            as_numer_denom(divx, outArg(divx_num), outArg(divx_den));
+            // if (eq(*divx_den, *one)) {
+            //     curr_num = add(mul(arg_num, divx), curr_num);
+            //     continue;
+            // } !! The below two 'general' statements cover this case
```

Comment:
Ah I see --- so the if statement is a special case, but it turns out that your general case already handles the special case. Is that right? Why don't you summarize what you just wrote in a small comment in the code to make this clear?